clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -0,0 +1,182 @@
+
+  public GrpcTransportClient(ClientConfig clientConfig) {
+    serverGrpcChannels = new VeniceConcurrentHashMap<>();
+    stubCache = new VeniceConcurrentHashMap<>();
+    nettyAddressToGrpcAddressMap = clientConfig.getNettyServerToGrpcAddressMap();
+    this.r2TransportClient = new R2TransportClient(clientConfig.getR2Client());
+  }
+
+  protected ManagedChannel getChannel(String serverAddress) {
+    if (!nettyAddressToGrpcAddressMap.containsKey(serverAddress)) {
+      throw new VeniceException("No grpc server found for address: " + serverAddress);
+    }
+
+    return serverGrpcChannels.computeIfAbsent(
+        serverAddress,
+        k -> ManagedChannelBuilder.forTarget(nettyAddressToGrpcAddressMap.get(k)).usePlaintext().build());
+  }
+
+  protected VeniceReadServiceGrpc.VeniceReadServiceStub getStub(ManagedChannel channel) {
+    return stubCache.computeIfAbsent(channel, VeniceReadServiceGrpc::newStub);
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
+    String[] requestParts = requestPath.split("/");
+    if (!requestParts[3].equals(STORAGE_ACTION)) {
+      LOGGER.debug(
+          "performing unsupported gRPC transport client action ({}), passing request to R2 client",
+          requestParts[3]);
+      return r2TransportClient.get(requestPath, headers);
+    }
+
+    ManagedChannel channel = getChannel(requestParts[2]);
+
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
+        .setPartition(Integer.parseInt(requestParts[5]))
+        .setKeyString(requestParts[6])
+        .setIsBatchRequest(false)
+        .build();
+
+    VeniceReadServiceGrpc.VeniceReadServiceStub clientStub = getStub(channel);
+    GrpcTransportClientCallback callback = new GrpcTransportClientCallback(clientStub, request);
+
+    return callback.get();
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> post(
+      String requestPath,
+      Map<String, String> headers,
+      byte[] requestBody) {
+    String[] requestParts = requestPath.split("/");
+    if (!requestParts[2].equals(STORAGE_ACTION)) {
+      LOGGER.debug(
+          "performing unsupported gRPC transport client action ({}), passing request to R2 client",
+          requestParts[2]);
+      return r2TransportClient.post(requestPath, headers, requestBody);
+    }
+
+    ManagedChannel channel = getChannel(requestParts[2]);
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
+        .setKeyBytes(ByteString.copyFrom(requestBody))
+        .setIsBatchRequest(true)
+        .build();
+
+    VeniceReadServiceGrpc.VeniceReadServiceStub clientStub = getStub(channel);
+    GrpcTransportClientCallback callback = new GrpcTransportClientCallback(clientStub, request);
+    return callback.post();
+  }
+
+  @Override
+  public void close() {
```

Comment:
It is good to call `r2TransportClient.close()` here even that function is doing nothing now.